### DISCUSSION TITLE
Real world examples of DefaultConnectionCallback and DefaultChannelCallback

### DISCUSSION THREAD

**Discussion by quantum-booty** [gftea/amqprs#53] [General]
I'm pretty new to rabbit and I'm not sure what the purposes of connection and channel callbacks are.
The doc string suggests the DefaultConnectionCallback and DefaultChannelCallback are for demo and debugging purposes only, so are there any examples showing how they should be implemented in production?

**Comment by michaelklishin** [comment]
[Connections](https://rabbitmq.com/connections.html) and [channels](https://rabbitmq.com/channels.html) expose various events for applications to use.

For example, other clients expose events such as

 * Connection failures
 * Channel errors
 * Successful consumer registration
 * When a connection is [blocked due to a resource alarm](https://www.rabbitmq.com/connection-blocked.html) or unblocked
 * [Consumer cancellation notifications](https://www.rabbitmq.com/consumer-cancel.html)

and so on. Many applications won't care about most of them but the more rigorous error handling you need, the more such events you'd want to handle and at least log.

According to [callbacks.rs](https://github.com/gftea/amqprs/blob/main/amqprs/src/api/callbacks.rs), this is done via callback implementations in this client. So this is very similar in spirit to the [Java client](https://www.rabbitmq.com/api-guide.html), which is fairly extensively documented.
